nav2_costmap_2d/plugins/denoise_layer.cpp

```diff
@@ -0,0 +1,185 @@
+  declareParameter("minimal_group_size", rclcpp::ParameterValue(2));
+  // Pixels connectivity type
+  declareParameter("group_connectivity_type", rclcpp::ParameterValue(8));
+
+  const auto node = node_.lock();
+
+  if (!node) {
+    throw std::runtime_error("DenoiseLayer::onInitialize: Failed to lock node");
+  }
+  node->get_parameter(name_ + "." + "enabled", enabled_);
+
+  auto getInt = [&](const std::string & parameter_name) {
+      int param{};
+      node->get_parameter(name_ + "." + parameter_name, param);
+      return param;
+    };
+
+  const int minimal_group_size_param = getInt("minimal_group_size");
+
+  if (minimal_group_size_param <= 1) {
+    RCLCPP_WARN(
+      logger_,
+      "DenoiseLayer::onInitialize(): param minimal_group_size: %i."
+      " A value of 1 or less means that all map cells will be left as they are.",
+      minimal_group_size_param);
+    minimal_group_size_ = 1;
+  } else {
+    minimal_group_size_ = static_cast<size_t>(minimal_group_size_param);
+  }
+
+  const int group_connectivity_type_param = getInt("group_connectivity_type");
+
+  if (group_connectivity_type_param == 4) {
+    group_connectivity_type_ = ConnectivityType::Way4;
+  } else {
+    group_connectivity_type_ = ConnectivityType::Way8;
+
+    if (group_connectivity_type_param != 8) {
+      RCLCPP_WARN(
+        logger_, "DenoiseLayer::onInitialize(): param group_connectivity_type: %i."
+        " Possible values are  4 (neighbors pixels are connected horizontally and vertically) "
+        "or 8 (neighbors pixels are connected horizontally, vertically and diagonally)."
+        "The default value 8 will be used",
+        group_connectivity_type_param);
+    }
+  }
+  current_ = true;
+}
+
+void
+DenoiseLayer::reset()
+{
+  current_ = false;
+}
+
+bool
+DenoiseLayer::isClearable()
+{
+  return false;
+}
+
+void
+DenoiseLayer::updateBounds(
+  double /*robot_x*/, double /*robot_y*/, double /*robot_yaw*/,
+  double * /*min_x*/, double * /*min_y*/,
+  double * /*max_x*/, double * /*max_y*/) {}
+
+void
+DenoiseLayer::updateCosts(
+  nav2_costmap_2d::Costmap2D & master_grid, int min_x, int min_y, int max_x, int max_y)
+{
+  if (!enabled_) {
+    return;
+  }
+
+  if (min_x >= max_x || min_y >= max_y) {
+    return;
+  }
+
+  // wrap roi_image over existing costmap2d buffer
+  unsigned char * master_array = master_grid.getCharMap();
+  const int step = static_cast<int>(master_grid.getSizeInCellsX());
+
+  const size_t width = max_x - min_x;
+  const size_t height = max_y - min_y;
+  Image<uint8_t> roi_image(height, width, master_array + min_x, step);
+
+  try {
+    denoise(roi_image);
+  } catch (std::exception & ex) {
+    RCLCPP_ERROR(logger_, (std::string("Inner error: ") + ex.what()).c_str());
+  }
+
+  current_ = true;
+}
+
+void
+DenoiseLayer::denoise(Image<uint8_t> & image) const
+{
+  if (image.empty()) {
+    return;
+  }
+
+  if (minimal_group_size_ <= 1) {
+    return;  // A smaller group cannot exist. No one pixel will be changed
```

Comment:
I don’t mind, but just in case, I pay attention: a warning will already [be shown](https://github.com/ros-planning/navigation2/pull/2567/files/2222e25e82258a18839ae314ee516f6257ca3e8a#diff-102f2d1fcbf93fc1b8c041a37d7dbd6f5359c6f9616d28ef716ccffc9e5d5474R57) at the initialization stage (and, in fact, `minimal_group_size_` is always at least 1, I left the `<=` check for generality, in case the code changes in the future).

Reply:
Agree, this will be second warning, so it is not needed. Closing item.